landscape/client/snap_utils.py

```diff
@@ -0,0 +1,44 @@
+import yaml
+
+from landscape.client import snap_http
+from landscape.client.snap_http import SnapdHttpException
+
+
+def parse_assertion(assertion: str):
+    """Parse an assertion into key-value pairs."""
+    headers, signature = assertion.split("\n\n")
+    parsed = {}
+    parsed.update(yaml.safe_load(headers))
+    parsed["signature"] = signature
+    return parsed
+
+
+def get_assertions(assertion_type: str):
+    """Get and parse assertions."""
+    try:
+        response = snap_http.get_assertions(assertion_type)
+    except SnapdHttpException:
+        return
+
+    # the snapd API returns multiple assertions as a stream of
+    # bytes separated by double newlines, something like this:
+    # <assertion1-headers>: <value>
+    # <assertion1-headers>: <value>
+    #
+    # signature
+    #
+    # <assertion2-headers>: <value>
+    # <assertion2-headers>: <value>
+    #
+    # signature
+
+    # extract the assertion headers + their signatures as separate assertions
+    sections = response.result.decode().split("\n\n")
+    raw_assertions = [
+        "\n\n".join(sections[i : i + 2]) for i in range(0, len(sections), 2)
```

Comment:
This split/re-join could be restructured to be more immediately understandable.

Let's think of it as I/O: you have a bytestream with alternating header groups and signatures separated by `'\n\n'`. These are pairs: `(headers, signature)`. Seems like a perfect usecase for `itertools.bulk`. Too bad that's py3.12 only. Here's an alt:

```python3
sections = response.result.decode().split("\n\n")
parsed = []
rest = sections
while rest:
    headers, signature, *rest = rest
    parsed.append(parse_assertion(headers, signature))

return parsed
```

Let me know what you think.

Reply:
Updated to use the more readable code with a slight modification to handle cases where there are no assertions retrieved from the API (`response.result = b""` -> `response.result.decode().split("\n\n")
 = [""]` ).